OpenRA.Mods.Common/Pathfinder/PathSearch.cs

```diff
@@ -73,13 +71,19 @@ static CellInfoLayerPool LayerPoolForWorld(World world)
 			heuristic = heuristic ?? DefaultCostEstimator(locomotor, target);
 			var search = new PathSearch(graph, heuristic, heuristicWeightPercentage, loc => loc == target, recorder);
 
-			foreach (var sl in froms)
-				if (world.Map.Contains(sl))
-					search.AddInitialCell(sl, customCost);
+			AddInitialCells(world, locomotor, froms, customCost, search);
 
 			return search;
 		}
 
+		static void AddInitialCells(World world, Locomotor locomotor, IEnumerable<CPos> froms, Func<CPos, int> customCost, PathSearch search)
+		{
+			var customMovementLayers = world.GetCustomMovementLayers();
+			foreach (var sl in froms)
+				if (world.Map.Contains(sl) && (sl.Layer == 0 || customMovementLayers[sl.Layer].EnabledForLocomotor(locomotor.Info)))
```

Comment:
I don't know if `sl.Layer == 0` is correct, because in TS (like in #19063) the cell layer on map cell can be > 0, and when actor goes subterrain it may right at 0 layer.

At this time the actor dies and `SpawnActorOnDeath`, the newly spawned actor will still pathfind this invalid cell.

Reply:
The layer of a cell is different from the height of the cell. Layer 0 is used for the ground layer, which all units have access to. Other numbers refer to a custom movement layer. The actual number is meaningless, it just has to be a unique identifier. The custom layers are used when we want the unit to be on the map but it's not on the ground layer.

Currently we have tunnels (1), elevated bridges (4), subterranean (2) and jumpjet (3). When entering a tunnel or crossing over an elevated bridge you'll use those layers. Your cell might live at the same x,y location as the ground layer. But because you're on a different layer we can tell that you're in the tunnel (rather than on the ground above it) or that you're crossing over the bridge (rather than crossing on the ground under it).

The subterranean and jumpjet layers are interesting in that only certain units can use them. When a unit is travelling underground or via jumpjet it will be using those layers instead.

Units can move between the ground layer and one of the custom movement layers. This layer switches instantly because it's just an identifier and not something like height which could change over time. As soon as you step onto an elevated bridge your layer switches from 0 to 4. As soon as it goes subterranean the layer switches from 0 to 2.

The height of a cell is relevant only on the ground layer. Cell height causes ramps and cliffs because cells can be higher than one another.

With that background out of the way, we can see why this check is correct. Layer 0 is checking if you're on the ground layer, which is allowed because all units have access to it. Any other layer requires us to check the custom movement layer and ask whether this unit is allowed to use it.

If your actor dies on the ground layer (0) then when the spawned actor will be able to move according to the usual rules. If it dies on a custom layer (e.g. 2 in the bug example) then the spawned unit will also spawn on that layer. This fix means we won't crash. The unit itself will be unable to move because it's on a custom movement layer. In the same way if it spawned on impassable terrain it would also be stuck.

So you still have the conceptual question of what to do with traits than can spawn actors if the spawned actor has different movement rules. If a subterrenean unit spawns a regular actor it's going to be stuck underground. If a naval unit spawns an infantry that infantry would be stuck on the water - and that's all on the ground layer, layer 0.

This PR means the scenario with the custom movement layer will no longer crash. Instead it will now be stuck just like the water example.

Reply:
It might be good to mention this somewhere, perhaps https://github.com/OpenRA/OpenRA/wiki/World-Coordinate-System. When looking for existing explanations I only found https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/World/ActorMap.cs#L669-L676

Reply:
Thanks  for the explanation